posts/process-csvs-from-amazon-s3-using-apache-flink-jhipster-and-kubernetes.md

```diff
@@ -0,0 +1,187 @@
+        .equalTo("state")
+        .map(new MapToCovidFunction())
+        .sortPartition(new IdSelectorDeath(), Order.DESCENDING)
+        .collect();
+}
+```
+
+Reading a CSV is pretty simple and Flink will do all the parsing job for you. Manipulating each `DataSource` is also pretty easy and operations like filtering/mapping can done directly on the data. The COVID-19 cases/deaths data source is filtered to only keep the values for the date `01/04/2021` since we are using the vaccine data for this date too.
+
+Once we have all the correct data, they all can be joined using the `state` as the joining key. `IdSelectorStateCode` takes care of converting a state name to its code so it correctly matches. After the joins, everything is mapped to one final POJO `VaccineStatsVM` containing the fields below:
+
+```java
+private String state;
+private String moderna;
+private String pfizer;
+private Long cases;
+private Long death;
+```
+
+Finally, the list is sorted by the number of `death` and the result is collected into a regular `List`. Because of the `collect()`, there is no need to call `env.execute()` which is usually required since all Flink operations are executed lazily.
+
+## Result and future work
+
+To trigger the job, we need to create an API endpoint in the JHipster application so it is available in the Swagger UI:
+
+```java
+@RestController
+@RequestMapping("/api")
+public class CovidResource {
+
+    @GetMapping("/covid")
+    public ResponseEntity<List<VaccineStatsVM>> getVaccineStats() throws Exception {
+        return ResponseEntity.ok(FlinkJob.run());
+    }
+
+}
+```
+
+Now we can compile and start our JHipster application:
+
+```bash
+# Compile the application and the job
+./mvnw clean verify
+
+# Start the application
+./mvnw
+```
+
+Browse to [http://localhost:8080](http://localhost:8080) and then login using the credentials `admin/admin`. Using the top menu, navigate to `Administration -> API` and you should see the resource `covid-resource`:
+
+![API](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-flink/images/2021/01/flink-2.png)
+
+After clicking on `Execute`, the job will be triggered and you will see the result after few seconds:
+
+![Result](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-flink/images/2021/01/flink-3.png)
+
+Future work would be to build a small UI to show the results in a chart rather than just a json. Also, it should be possible to pick the date and the field to sort on.
+
+# Job manager’s web UI
+
+What is pretty cool about Flink is that you have access to a pretty [neat dashboard](http://localhost:8081/). The screenshot below shows all the different stages on the job we just ran:
+
+![Dashboard](https://raw.githubusercontent.com/Falydoor/blog-usa/blog-flink/images/2021/01/flink-4.png)
+
+We can see all the tasks executed and their duration, also, the diagram is all interactive which makes things very easy to navigate in case of a complex job.
```

Comment:
```suggestion
We can see all the tasks executed and their duration. Also, the diagram is all interactive which makes things very easy to navigate in case of a complex job.
```